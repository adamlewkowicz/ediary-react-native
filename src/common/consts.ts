export const PRODUCT_UNITS = ['g', 'mg', 'kg', 'ml', 'l'] as const;
export const BASE_MACRO_ELEMENTS = ['carbs', 'prots', 'fats'] as const
export const MACRO_ELEMENTS = [...BASE_MACRO_ELEMENTS, 'kcal'] as const;