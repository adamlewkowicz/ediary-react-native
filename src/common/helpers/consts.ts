import { MacroElements } from '../../types';
import { DiaryTemplate } from '../../store/reducers/diary';

export const baseMacro: MacroElements = {
  carbs: 0,
  prots: 0,
  fats: 0,
  kcal: 0
}

export const defaultTemplates = [
  {
    id: 1,
    name: 'Śniadanie I',
    time: '08:30:00',
  },
  {
    id: 2,
    name: 'Śniadanie II',
    time: '11:00:00',
  },
  {
    id: 3,
    name: 'Obiad',
    time: '13:00:00',
  },
  {
    id: 4,
    name: 'Podwieczorek',
    time: '17:00:00',
  },
  {
    id: 5,
    name: 'Kolacja',
    time: '19:00:00',
  },
] as any as DiaryTemplate[];