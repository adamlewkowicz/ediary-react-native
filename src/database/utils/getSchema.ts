import { getConnection, Connection } from 'typeorm';
import sqlFormatter from 'sql-formatter';

export async function getDatabaseSchema(
  connection: Connection = getConnection(),
  metaTableNames: string[] = ['sqlite_sequence', 'migrations']
): Promise<ExtendedTableSchema[]> {
  const tableSchema: TableSchema[] = await connection.query(
    `SELECT * FROM sqlite_master WHERE type = "table";`
  );
  const normalizedSchema: ExtendedTableSchema[] = tableSchema
    .filter(data => !metaTableNames.includes(data.tbl_name))
    .map(data => ({
      ...data,
      prettySql: sqlFormatter.format(data.sql)
    }))
    .sort((a, b) => a.tbl_name > b.tbl_name ? 1 : -1);
    
  return normalizedSchema;
}

interface ExtendedTableSchema extends TableSchema {
  prettySql: string
}
interface TableSchema {
  name: string
  rootpage: number
  sql: string
  tbl_name: string
  type: 'table'
}