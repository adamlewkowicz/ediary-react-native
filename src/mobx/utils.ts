import { Training, ExerciseSet, Exercise } from '../database/entities';
import { ExerciseSetState, ExerciseState, TrainingState } from './GymTrainingStore';

export const normalizeExerciseSet = (exerciseSet: ExerciseSet): ExerciseSetState => ({
  ...exerciseSet,
  duration: 0,
  isRest: false,
  restTime: 20,
  restDuration: 0,
  state: 'unfinished',
});

export const normalizeExercise = (exercise: Exercise) => {
  const { sets = [], ..._exercise } = exercise;

  const normalizedExerciseSets = sets.map(normalizeExerciseSet);

  const normalizedExercise: ExerciseState = {
    ..._exercise,
    duration: 0,
    setIds: normalizedExerciseSets.map(set => set.id),
  }

  return {
    exercise: normalizedExercise,
    exerciseSets: normalizedExerciseSets,
  }
}

const normalizeExercises = (exercises: Exercise[]) => {
  return exercises.reduce<NormalizedExercises>((normalized, _exercise) => {
    const { exercise, exerciseSets } = normalizeExercise(_exercise);

    return {
      exercises: [...normalized.exercises, exercise],
      exerciseSets: [...normalized.exerciseSets, ...exerciseSets]
    };
  }, { exercises: [], exerciseSets: [] });
}

export function normalizeTrainings(payload: Training[]) {
  return payload.reduce<NormalizedTrainings>((normalized, training) => {
    const { exercises = [], ...restTraining } = training;

    const result = normalizeExercises(exercises);

    const normalizedTraining: TrainingState = {
      ...restTraining,
      exerciseIds: exercises.map(exercise => exercise.id),
      isPaused: false,
      isActive: false,
    }

    return {
      trainings: [...normalized.trainings, normalizedTraining],
      exercises: [...normalized.exercises, ...result.exercises],
      exerciseSets: [...normalized.exerciseSets, ...result.exerciseSets],
    };
  }, { trainings: [], exercises: [], exerciseSets: [] });
}

interface NormalizedExercises {
  exercises: ExerciseState[]
  exerciseSets: ExerciseSetState[]
}

interface NormalizedTrainings extends NormalizedExercises  {
  trainings: TrainingState[]
}