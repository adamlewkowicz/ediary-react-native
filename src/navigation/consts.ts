
export const APP_ROUTE = {
  AppLoading: 'AppLoading',
  ProfileCreate: 'ProfileCreate',
  MainStack: 'MainStack',
    NutritionStack: 'NutritionStack',
      NutritionHome: 'NutritionHome',
      ProductCreate: 'ProductCreate',
      ProductFind: 'ProductFind',
      BarcodeScan: 'BarcodeScan',
      ProductPreview: 'ProductPreview',
    DiarySummary: 'DiarySummary',
} as const;