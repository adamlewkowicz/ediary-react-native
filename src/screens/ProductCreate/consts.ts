
export const NUTRITION_INPUTS = [
  {
    title: 'Węglowodany',
    property: 'carbs',
    nextRef: 'prots'
  },
  {
    title: 'Białka',
    property: 'prots',
    nextRef: 'fats'
  },
  {
    title: 'Tłuszcze',
    property: 'fats',
    nextRef: 'kcal'
  },
  {
    title: 'Kalorie',
    property: 'kcal',
    nextRef: 'barcode'
  }
] as const;

export enum PORTION_TITLE {
  '100g' = 'Opakowanie zawiera',
  'portion' = 'Porcja zawiera',
  'package' = 'Opakowanie zawiera'
};