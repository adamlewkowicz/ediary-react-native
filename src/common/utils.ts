import dayjs from 'dayjs';
import { DateDay, UnitType } from '../types';
import { UNIT_TYPES } from './consts';

let timeout: NodeJS.Timeout;

export const debounce_ = () => {
  let timeout: NodeJS.Timeout;
  return (fn: () => void, delay = 250) => {
    clearTimeout(timeout);
    timeout = setTimeout(fn, delay);
  }
}

export const debounce = (fn: () => void, delay = 150) => {
  clearTimeout(timeout);
  timeout = setTimeout(fn, delay);
}

export const round = (value: number, scale = 10) => Math.round(value * scale) / scale;

export const toLocaleString = (value: number) => new Intl.NumberFormat('pl-PL').format(value);

export const mapAsyncSequence = async <T, R>(
  items: T[],
  callback: (item: T) => Promise<R>
) => {
  const result: R[] = [];

  for (const item of items) {
    const resultItem = await callback(item);
    result.push(resultItem);
  }

  return result;
}

export function getDayFromDate(
  date: dayjs.ConfigType
): DateDay {
  const dateDay: DateDay = dayjs(date).format('YYYY-MM-DD') as any;
  return dateDay;
}


export function getValAndUnitFromString(value: string): {
  value: number | null
  unit: UnitType | null
} {
  const unit = UNIT_TYPES.find(unit => value.includes(unit));
  const parsedValue = parseFloat(
    value.replace(/,/, '.')
  );
  const result = Number.isNaN(parsedValue) ? null : parsedValue; 

  return {
    unit: unit || null,
    value: result || null
  }
}

export function sortByMostAccurateName(
  phrase: string
) {
  const phraseLowered = phrase.toLowerCase();

  return function<T extends SortByName>(
    a: T,
    b: T
  ): -1 | 0 | 1 {
    const aLowered = a.name.toLowerCase();
    const bLowered = b.name.toLowerCase();
  
    const aIndex = aLowered.indexOf(phraseLowered);
    const bIndex = bLowered.indexOf(phraseLowered);
  
    const notFoundIndex = -1;
    const orderByShorter = aLowered.length > bLowered.length ? 1 : -1;
  
    if (aIndex === notFoundIndex && bIndex === notFoundIndex) {
      return -1;
    }
    if (aIndex === notFoundIndex) {
      return 1;
    }
    if (bIndex === notFoundIndex) {
      return -1;
    }
  
    if (aLowered.length === bLowered.length) {
      return 0;
    }
  
    return orderByShorter;
  }
}

interface SortByName {
  name: string
}

export function getNumAndUnitFromString(value: string): {
  value: number | null
  unit: UnitType | null
} {
  const parseToNumber = (val: string): number | null => {
    const parsed = val.trim().replace(/,/, '.')
    const numbers = parsed.match(/[+-]?\d+(\.\d+)?/g);

    if (!numbers) {
      return null;
    }
    const result = parseFloat(numbers[numbers.length - 1]);

    if (Number.isNaN(result)) {
      return null;
    }
    return result;
  }

  const unit = UNIT_TYPES.find(unit => value.includes(` ${unit}`)) || null;
  const result = parseToNumber(value);

  return {
    value: result,
    unit
  }
}

export function filterByUniqueId<T extends { id: number | string }>(
  item: T,
  index: number,
  self: T[]
): boolean {
  const foundIndex = self.findIndex(anyItem => anyItem.id === item.id);
  return foundIndex === index;
}

export function parseNumber(
  value: string,
  maxNumber?: number,
  maxLength?: number,
): string {
  const trimmed = maxLength ? value.substring(0, maxLength) : value;
  const cleaned = trimmed.replace(/[^\d.-]/g, '');
  const parsed = Number(cleaned);

  if (maxNumber && parsed >= maxNumber) {
    return cleaned.slice(0, -1);
  }

  return cleaned;
}

export function findOrFail<T>(
  array: T[],
  matcher: (item: T, index: number, self: T[]) => boolean
) {
  const foundItem = array.find(matcher);
  if (!foundItem) {
    throw new Error(
      '[findOrFail] - Item could not be found for specified criteria'
    );
  }
  return foundItem;
}