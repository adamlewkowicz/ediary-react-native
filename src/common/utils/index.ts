/// <reference lib="dom" />
import dayjs from 'dayjs';
import { DateDay, UnitType, DateTime, MacroElements, BaseMacroElements, ObjectEntries } from '../../types';
import { UNIT_TYPES, DATE_TIME, DATE_DAY, MACRO_ELEMENTS, KCAL_IN_ONE_MACRO_GRAM } from '../consts';
import { LayoutAnimation } from 'react-native';

export const debounce = () => {
  let timeout: NodeJS.Timeout;
  return (callback: () => void, delay = 250) => {
    clearTimeout(timeout);
    timeout = setTimeout(callback, delay);
  }
}

export const round = (value: number, scale = 10) => Math.round(value * scale) / scale;

export const toLocaleString = (value: number) => new Intl.NumberFormat('pl-PL').format(value);

/**
 * Maps over an array and runs promise returned from a callback in a sequence.
 * @param items array of elements to iterate on
 * @param callback callback that is executed on each iteration
 */
export const mapAsyncSequence = async <T, R>(
  items: T[],
  callback: (item: T) => Promise<R>
) => {
  const result: R[] = [];

  for (const item of items) {
    const resultItem = await callback(item);
    result.push(resultItem);
  }

  return result;
}

export function getDayFromDate(
  date: dayjs.ConfigType
): DateDay {
  const dateDay: DateDay = dayjs(date).format(DATE_DAY) as any;
  return dateDay;
}

export function getTimeFromDate(
  date: dayjs.ConfigType
) {
  const dateTime: DateTime = dayjs(date).format(DATE_TIME) as any;
  return dateTime;
}

export function getValAndUnitFromString(value: string): {
  value: number | null
  unit: UnitType | null
} {
  const unit = UNIT_TYPES.find(unit => value.includes(unit));
  const parsedValue = parseFloat(
    value.replace(/,/, '.')
  );
  const result = Number.isNaN(parsedValue) ? null : parsedValue; 

  return {
    unit: unit || null,
    value: result || null
  }
}

export function sortByMostAccurateName(
  phrase: string
) {
  const phraseLowered = phrase.toLowerCase();

  return function<T extends { name: string }>(
    a: T,
    b: T
  ): -1 | 0 | 1 {
    const aNameLowered = a.name.toLowerCase();
    const bNameLowered = b.name.toLowerCase();
  
    const aIndex = aNameLowered.indexOf(phraseLowered);
    const bIndex = bNameLowered.indexOf(phraseLowered);
  
    const notFoundIndex = -1;
    const orderByShorter = aNameLowered.length > bNameLowered.length ? 1 : -1;

    const aNameHasNotBeenFound = aIndex === notFoundIndex;
    const bNameHasNotBeenFound = bIndex === notFoundIndex;
  
    if (aNameHasNotBeenFound && bNameHasNotBeenFound) {
      return -1;
    }
    if (aNameHasNotBeenFound) {
      return 1;
    }
    if (bNameHasNotBeenFound) {
      return -1;
    }
  
    if (aNameLowered.length === bNameLowered.length) {
      return 0;
    }
  
    return orderByShorter;
  }
}

export function getNumAndUnitFromString(value: string): {
  value: number | null
  unit: UnitType | null
} {
  const parseToNumber = (val: string): number | null => {
    const parsed = val.trim().replace(/,/, '.')
    const numbers = parsed.match(/[+-]?\d+(\.\d+)?/g);

    if (!numbers) {
      return null;
    }
    const result = parseFloat(numbers[numbers.length - 1]);

    if (Number.isNaN(result)) {
      return null;
    }
    return result;
  }

  const unit = UNIT_TYPES.find(unit => value.includes(` ${unit}`)) || null;
  const result = parseToNumber(value);

  return {
    value: result,
    unit
  }
}

export const filterByUniqueProperty = <
  E extends object,
  A extends E[]
>(property: keyof E) => (element: E, index: number, self: A): boolean => {
  const foundIndex = self.findIndex(anyElement =>
    anyElement[property] === element[property]
  );
  return foundIndex === index;
}

export const filterByUniqueId = filterByUniqueProperty('id');

export function parseNumber(
  value: string,
  maxNumber?: number,
  maxLength?: number,
): string {
  const trimmed = maxLength ? value.substring(0, maxLength) : value;
  const cleaned = trimmed.replace(/[^\d.-]/g, '');
  const parsed = Number(cleaned);

  if (maxNumber && parsed >= maxNumber) {
    return cleaned.slice(0, -1);
  }

  return cleaned;
}

export function findOrFail<T>(
  array: T[],
  matcher: (item: T, index: number, self: T[]) => boolean
): T {
  const foundItem = array.find(matcher);
  if (!foundItem) {
    throw new Error(
      '[findOrFail] - No item could be found for specified criteria'
    );
  }
  return foundItem;
}

const createArrayOfLength = <T>(
  length: number,
  callback: (index: number) => T
): T[] => Array.from({ length }, (_, index) => callback(index));

export const fillArrayWithinRange = (
  { from, to }: { from: number; to: number }
): number[] => {
  const zeroBasedCountFill = 1;
  const length = to - from + zeroBasedCountFill;
  return createArrayOfLength(length, index => index + from);
}

export const fetchify = async <T>(
  input: RequestInfo,
  init: RequestInit = {},
  controller = new AbortController()
): Promise<T> => {
  const { signal } = controller;

  const response = await fetch(input, { signal, ...init });
  const json: T = await response.json();

  return json;
}

export const objectMap = <
  T extends object,
  Property extends keyof T = keyof T,
  Value = T[Property]
>(
  obj: T,
  callback: (property: Property, value: Value) => Value
): T => {
  return Object.fromEntries(
    Object
      .entries(obj)
      .map(([property, value]) => [property, callback(property as Property, value)])
  ) as T;
}

const objectEntries: ObjectEntries = Object.entries;

export const calculateCaloriesByMacro = (
  macro: BaseMacroElements
): number => {
  return objectEntries(macro)
    .reduce(
      (kcal, [macroName, macroQuantity]) => kcal += macroQuantity * KCAL_IN_ONE_MACRO_GRAM[macroName],
      0
    );
}

export const layoutAnimateEase = (onAnimationDidEnd?: () => void) => {
  LayoutAnimation.configureNext(
    LayoutAnimation.Presets.easeInEaseOut,
    onAnimationDidEnd
  );
}

export const calculatePercentage = (portion: number, total: number): number => {
  if (total === 0) return 0;
  return Math.floor(portion / total * 100);
}

export const calculateMacroPerQuantity = <T extends MacroElements>(
  macroValuesPerHundredQuantity: T,
  quantity: number
): T => objectMap(
  macroValuesPerHundredQuantity,
  (_, macroValue: number) => round(macroValue * quantity / 100, 100)
);

export const calculateMacroPercentages = <T extends BaseMacroElements>(
  macroValues: T
): BaseMacroElements => {
  const macroSum = macroValues.carbs + macroValues.prots + macroValues.fats;

  const carbs = calculatePercentage(macroValues.carbs, macroSum);
  const prots = calculatePercentage(macroValues.prots, macroSum);
  const fats = calculatePercentage(macroValues.fats, macroSum);

  return { carbs, prots, fats };
}

const calculateMacroValueNeed = (
  eaten: number,
  needed: number
): MacroNeed => ({
  eaten,
  needed,
  left: needed - eaten,
  percentage: calculatePercentage(eaten, needed),
});

export const calculateMacroNeeds = <T extends object>(
  macroValuesEaten: T,
  macroValuesNeeded: T
): { [key: string]: MacroNeed } => objectMap(
  macroValuesEaten,
  (macroName, macroValueEaten) => {
    // TODO: refactor types
    const macroValueNeeded = macroValuesNeeded[macroName] as any as number;

    const macroValueNeed = calculateMacroValueNeed(
      macroValueEaten,
      macroValueNeeded
    );

    return macroValueNeed;
  }
);

type MacroNeed = {
  eaten: number
  needed: number
  left: number,
  percentage: number
}

const calculateObjectsValueSum = <T extends { [key: string]: number }>(
  objects: T[],
  initialObj: T = objects[0],
): T => {
  return objects.reduce<T>((acc, current) => {
    const next = { ...acc };

    for (const key in next) {
      // @ts-ignore
      next[key] += current[key];
    }

    return next;
  }, { ...initialObj });
}