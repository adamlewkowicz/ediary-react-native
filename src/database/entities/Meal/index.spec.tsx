import { Meal } from '.';
import { getDayFromDate } from '../../../common/utils';
import dayjs from 'dayjs';
import { DATE_FORMAT } from '../../../common/consts';

test('Meal.getMacroHistory - groups meals by day and returns summed macro values sorted ascendingly by day', async () => {
  const startDateMock = '2019-01-07';
  const endDateMock = '2019-01-01';
  const daysInWeek = 7;

  const startDate = new Date(startDateMock);
  const endDate = new Date(endDateMock);

  const startDateString = dayjs(startDate).format(DATE_FORMAT);
  const endDateString = dayjs(endDate).format(DATE_FORMAT);

  const mealA1 = await Meal.save({ name: 'A1', date: startDateString, macro: { carbs: 40, kcal: 10 }});
  const mealA2 = await Meal.save({ name: 'A2', date: startDateString, macro: { carbs: 40, kcal: 10 }});
  const mealB1 = await Meal.save({ name: 'B1', date: endDateString, macro: { carbs: 81, kcal: 10 }});

  const result = await Meal.getMacroHistory(getDayFromDate(startDate));
  const [firstRecord] = result;
  const lastRecord = result[result.length - 1];

  expect(result.length).toEqual(daysInWeek);
  expect(lastRecord.day).toEqual(startDateMock);
  expect(lastRecord.carbs).toEqual(mealA1.macro.carbs + mealA2.macro.carbs);
  expect(lastRecord.prots).toEqual(0);
  expect(firstRecord.carbs).toEqual(mealB1.macro.carbs);
});

test('Meal.getMacroSummary - returns macro history and calculates average values', async () => {
  const dayAMock = '2019-01-06';
  const dayBMock = '2019-01-03';

  const dateA = new Date(dayAMock);
  const dateB = new Date(dayBMock);

  const dateAString = dayjs(dateA).format(DATE_FORMAT);
  const dateBString = dayjs(dateB).format(DATE_FORMAT);

  await Meal.save({ name: 'A1', date: dateAString, macro: { carbs: 40, kcal: 10 }});
  await Meal.save({ name: 'A2', date: dateAString, macro: { carbs: 40, kcal: 10 }});
  await Meal.save({ name: 'B1', date: dateBString, macro: { carbs: 40, kcal: 10 }});

  const result = await Meal.getMacroSummary(getDayFromDate(dateA));

  expect(result).toMatchSnapshot();
});