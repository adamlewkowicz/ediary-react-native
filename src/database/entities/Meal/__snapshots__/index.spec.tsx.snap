// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meal.getMacroSummary - returns macro history and calculates average values 1`] = `
Object {
  "average": Object {
    "carbs": 17,
    "fats": 0,
    "kcal": 4,
    "prots": 0,
  },
  "data": Array [
    Object {
      "carbs": 0,
      "day": "2018-12-31",
      "fats": 0,
      "kcal": 0,
      "prots": 0,
    },
    Object {
      "carbs": 0,
      "day": "2019-01-01",
      "fats": 0,
      "kcal": 0,
      "prots": 0,
    },
    Object {
      "carbs": 0,
      "day": "2019-01-02",
      "fats": 0,
      "kcal": 0,
      "prots": 0,
    },
    Object {
      "carbs": 40,
      "day": "2019-01-03",
      "fats": 0,
      "kcal": 10,
      "prots": 0,
    },
    Object {
      "carbs": 0,
      "day": "2019-01-04",
      "fats": 0,
      "kcal": 0,
      "prots": 0,
    },
    Object {
      "carbs": 0,
      "day": "2019-01-05",
      "fats": 0,
      "kcal": 0,
      "prots": 0,
    },
    Object {
      "carbs": 80,
      "day": "2019-01-06",
      "fats": 0,
      "kcal": 20,
      "prots": 0,
    },
  ],
}
`;
